### DISCUSSION TITLE
active sessions

### DISCUSSION THREAD

**Discussion by lakosm** [cybertec-postgresql/pg_timetable#428] [Q&A] [answered]
Hi,

I have now 2 workers. I initiated 10 multiple chains using self destruction and job_max_instances = 1.
After the 10 chain finished their run, I see 10 active pg_timetable sessions in pg_stat_activity and timetable.active_session contains 39 rows.
When will be the 10 active session idle? Because, I suspest, if I want schedule another 10 chains, may active sessions in pg_stat_activity will grow up to 20, which is not suitable for my 4 cpu db server.

Thanks in advance!

Regards,
Miklos

UPDATE: when I launced the next 10 chains, the active sessions were gone from pg_stat_activity.
But I would like to know, what the rule is regarding this

**Comment by pashagolub** [comment, marked as answer]
Hi, 

assuming when we are speaking about 2 workers we are talking about 2 `pg_timetable` instances.

Each `pg_timetable` instance can use max `--cron-workers + --interval-workers` parallel routines according to the [manual](https://pg-timetable.readthedocs.io/en/master/README.html#command-line-options).

In the worst scenario (highest load) we need:
- separate connection for each of routines, 
- separate connection for `Scheduler.retrieveChainsAndRun()`, 
- separate connection for `Scheduler.retrieveIntervalChainsAndRun()`
- and another connection for `LogHook.send()`. 
 
That means the max connection (session) number for one `pg_timetable` instance (worker) would be:
```
	max_conn = --cron-workers + --interval-workers + 3
```

In other words, if you want to limit the number of the database sessions you should decrease the `--cron-workers` and/or `--interval-workers` start parameters (default value 16 for both).

`job_max_instances` applies only to the chain. Since you have self-destructed chains it does nothing. This parameter allows to run the same chain in parallel but is limited to the `job_max_instances` value, e.g. I want to start every minute some chain to process smth but no more than 10 chains at once. When the old chain finishes system will start a new one, so, at any point in time, there will be no more than `job_max_instances` running chains.

I hope I explained clearly enough. :-)

**Reply by lakosm** [reply]
Hi,

Wow, thank you for your detailed explaination. Now, I see.
I had issues with the default null job_max_instances (pk violation) so this is why I set it to 1.

**Reply by pashagolub** [reply]
You are more than welcome. 

Strange, `timetable.chain.max_instances       INTEGER`... It can be `NULL`, which means `unlimited`. [More details](https://pg-timetable.readthedocs.io/en/master/components.html#table-timetable-chain)...

What version are you running?

We should probably add a statement about the `NULL` value to make things more clear.

**Reply by lakosm** [reply]
I am using pg_timetable_4.6.1 and postgresql 13.3
Yesterday I got some "Message": "duplicate key value violates unique constraint errors. After I modified the job_max_instances to 1, the error disappeared.
I regularly check timetable.log table for errors

**Comment by DaskalovBG** [comment]
Hi,

I am trying to utilize pgtimetable on a semi-loaded vanilla postgres. I have 65 procedures running independent from one another and in addition to this I have only 3 chains running 2 procedures each chain (need to honor some dependency).

My server have some 16 CPUs and 96GB of RAM. 

What I observe is as if my procedures keep piling up. I only have around 50 to 60 active sessions when I query the active_session table and whenever I check pg_stat_activity there is only 1 instance of a session form pgtimetable.

My goal is to run all these 65 procs on a minute basis (each proc takes between 5 and 15 seconds to finish). But I seem to wait for workers to free up to pick up the next run.

Max current workers is set to 900, session keep staying between 50 and 60.

Would you mind helping me orientate in the given situation ?

**Reply by pashagolub** [reply]
Hi,

to run 65 procedures simultaneously you'd need a 65 workers. Anyway, you might want to increase these parameters:

```
      --cron-workers=                              Number of parallel workers for scheduled chains (default: 16)
      --interval-workers=                          Number of parallel workers for interval chains (default: 16)
````
And, of course, please check Postgres `max_connections` [parameter](https://postgresqlco.nf/doc/en/param/max_connections/)

**Reply by DaskalovBG** [reply]
question, by workers you mean pg_timetable --clientanem=worker1 or different form of workers. 

At the moment I have 1 instance of pg_timetable started, with 

/config # cat config.yaml 
clientname: brave_worker
no-program-tasks: true
...
resource:
  cron-workers: 200
  interval-workers: 200
  chain-timeout: 0
  task-timeout: 0 

but still only 1 instance of pg_timestable.

Also in the respective database in pg_stat_activity there is only 1 active (doing work) connection with name brave_worker, the rest are idle in transaction.

I found https://www.cybertec-postgresql.com/en/postgresql-exclusive-cron-jobs-using-pg_timetable-scheduler/ looks like I need to spawn more pg_timetable --clientname=workerX

**Reply by pashagolub** [reply]
No, by workers I mean internal workers (threads) inside pg_timetable instance.

Ok, so now I have a question. Show me the definition of your chains. I suppose you're running them in exclusive mode and if so, I'm wondering why do you want to run more pg_timetable instances. I mean, you want exclusive or parallel chains at the end of the day? :)

**Reply by DaskalovBG** [reply]
SELECT timetable.add_job(
    job_name            => 'my_job_name',
    job_schedule        => '0 0 * * *',
    job_command         => 'select my_etl_1()',
    job_parameters      => NULL,
    job_kind            => 'SQL'::timetable.command_kind,
    job_client_name     => NULL,
    job_max_instances   => 1,
    job_live            => TRUE,
    job_exclusive       => TRUE,
    job_self_destruct   => FALSE,
    job_ignore_errors   => FALSE
) as chain_id;

I have 61 of these.

I have 4 of these:

do $$
declare
    v_parent_id  bigint;
    v_task_id    bigint;
    v_chain_id   bigint;
    l_sqlstate   text;
    l_sqlerror   text;
begin
    -- create chain
    insert into timetable.chain (
        chain_id,
        chain_name,
        run_at,
        max_instances,
        live,
        self_destruct,
        exclusive_execution
    ) values (
        default,            -- chain_id,
        'my_specific_dependency',  -- chain_name
        '* * * * *',        -- run_at,
        1,                  -- max_instances,
        true,               -- live,
        false,              -- self_destruct,
        true                -- exclusive_execution,
    ) returning chain_id into v_chain_id;

    -- add parent task number 1
    insert into timetable.task (chain_id, task_order, command, ignore_error)
    values (v_chain_id, 1, 'select etl_1()', false)
    returning task_id into v_parent_id;

    -- add 1st child task number 2 dependent on parent task
    insert into timetable.task (chain_id, task_order, command, ignore_error)
    values (v_chain_id, 2, 'select etl_2()', false)
    returning task_id into v_task_id;

exception when others then get stacked diagnostics l_sqlstate = RETURNED_SQLSTATE, l_sqlerror = MESSAGE_TEXT;
    raise exception 'chain failed, %, %', l_sqlstate, l_sqlerror;

end;
$$ language plpgsql;


So the 61 are independent one from another (processing 61 different things) I need them to run whenever they finished running (ideally once a minute). 

The chain dependency I need to run specifically honoring the etl1 etl2 etl3 procs

**Reply by pashagolub** [reply]
Ok, we have an answer:


```
job_exclusive => TRUE,
```

This means you want to run every your chain exclusively. Means no other chains can run at this moment. Make it false and we're done :)

**Reply by DaskalovBG** [reply]
working on all job deffs to make it false, as we speak. 
Will keep you posted!
Thanks for the immense help!

**Reply by DaskalovBG** [reply]
questions:

- do I need to restart the pg_timetable instance ?
- how can I check if more tan 1 proc is being executed at once (I still see only 1 session of brave_worker in pg_stat_activity)

**Reply by pashagolub** [reply]
Nope, pg_timetable will catch modifications up.
You can check log output or `TABLE timetable.active_chain`

**Reply by DaskalovBG** [reply]
Interesting, by the looks of it, I have achieved what I wanted, but for the past hour only 60% of my procedures ran, as if pg_timetable never reached them, while procs which are in these 60% are executed a number of times withing the given period per minute. (I am running 900 workers and 900 interval-workers).

There is 1 procedure which takes around 30 minutes to complete, for that time noone else seems to be working (no other procs are executed). Task is added with job_exclusive       => FALSE, and FALSE -- exclusive_execution

Any ideas ?

**Reply by pashagolub** [reply]
Only one. You might want to restart pg_timetable because of piled up exclusive chains. New will be the normal but previous are still exclusive in queue

**Reply by DaskalovBG** [reply]
this thing went crazy!

Thank you for the immense support Pasha!

**Reply by pashagolub** [reply]
You're welcome! :)
